Permit initializers in library functions

tautschnig opened the pull request:
The goal of this series of commits is to reduce the number of `__CPROVER_...` static-lifetime objects that appear in every program irrespective of their use by C (model) library functions. This will limit verification overhead to those cases where they are actually used. This will also enable more costly objects (such as unbounded arrays) to be used by future library instrumentation.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.

If your PR fixes a bug, include the regression test(s) in the same commit as the bug fix. Else, keep commits small and orthogonal, possibly placing tests in commits of their own.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
Use the linker when adding library functions

This enables type checking of missing functions vs library-provided functions,
and also enables type sanitisation as done by the linker.

Co-authored-by: Peter Schrammel <peter.schrammel@diffblue.com>

tautschnig committed:
C library: move pipe() related definitions to library

With the support of building initialisers while adding the library we
can move library-only symbols to the library. This avoids polluting the
initialisation function for all cases that don't actually use
pipe-related library functions.

tautschnig committed:
CBMC library regression tests: ensure GNUC-only declarations are present

MSVC won't have the built-in declarations available, and will thus fail
type checking.

tautschnig committed:
enum-trace1 regression test: ensure inputs are sane

The test descriptions expect valid enum values, which we didn't actually
have any guarantee built into the test for.

chrisr-diffblue commented on src/ansi-c/library/pthread_lib.c:
```diff
@@ -290,12 +290,15 @@ inline int pthread_mutex_destroy(pthread_mutex_t *mutex)
-extern __CPROVER_thread_local unsigned long __CPROVER_next_thread_key;
+__CPROVER_bool __CPROVER_threads_exited[__CPROVER_constant_infinity_uint];
+__CPROVER_thread_local unsigned long __CPROVER_thread_id = 0;
+#if 0
```
Any particular reason for the `#if 0` rather than deletion? If theres a reason (other than "its a work in progress" :-) ) I'd suggest adding a comment to explain the reason.

tautschnig replied:
Comments added in all the places that have `#if 0`: thread key destructors are expensive. We'll have to figure out a viable solution at some point.

chrisr-diffblue replied:
Thanks, much appreciated :-) I always hate encountering an `#if 0` with not context as to why its there, whether it works if re-enabled, etc.

chrisr-diffblue commented on src/ansi-c/library/pthread_lib.c:
```diff
@@ -533,12 +540,17 @@ inline int pthread_rwlock_wrlock(pthread_rwlock_t *lock)
+#ifndef LIBRARY_CHECK
+__CPROVER_thread_local unsigned long __CPROVER_thread_id = 0;
+#endif
+#if 0
```
Ditto here and the other places in this file.

chrisr-diffblue reviewed (approved):
This seems reasonable and sensible to me.

remi-delmas-3000 commented on src/goto-programs/link_to_library.cpp:
```diff
@@ -25,24 +79,27 @@ Author: Daniel Kroening, kroening@kroening.com
-    &library)
+  const std::function<void(
+    const std::set<irep_idt> &,
+    const symbol_tablet &,
+    symbol_tablet &,
```
Can we document the two symbol table parameters and what they are supposed to be set to?

tautschnig replied:
Documentation added.

remi-delmas-3000 commented on src/goto-programs/link_to_library.cpp:
```diff
@@ -25,24 +79,27 @@ Author: Daniel Kroening, kroening@kroening.com
   // may depend on other library functions
 
   std::set<irep_idt> added_functions;
+  replace_symbolt::expr_mapt object_type_updates;
```
Could you explain the role of this new map with a comment ?

tautschnig replied:
Done.

remi-delmas-3000 commented on src/goto-programs/link_to_library.cpp:
```diff
@@ -59,30 +116,44 @@ void link_to_library(
       }
       else
-        missing_functions.insert(id);
+      {
+        changed = true;
+        added_functions.insert(id);
+
```
How about reformulating the control flow structure as
```c
if(body_available || already_added)
 continue;

// add the symbol
changed = true;
added_functions.insert(id);
...
 
```

tautschnig replied:
Could do, but then I'd have to touch more code in this PR.

remi-delmas-3000 commented on src/goto-programs/link_to_library.cpp:
```diff
@@ -11,9 +11,63 @@ Author: Daniel Kroening, kroening@kroening.com
+    goto_model.goto_functions.function_map.find(INITIALIZE_FUNCTION) !=
+    goto_model.goto_functions.function_map.end())
+  {
+    for(const auto &entry : library_model.symbol_table)
+    {
+      if(
```
if a same static variable is declared more than once in the library code, do we check that the initial value for the symbol is the same at each declaration ?

Eventhough each declaration follows the pattern below, depending on which one gets pulled first results could be different:

```c
/* FUNCTION : foo */
#ifndef blah_is_defined
#define blah_is_defined
static int blah = 12;
#endif

int foo() 
{
  // uses blah
}

/* FUNCTION : bar */
#ifndef blah_is_defined
#define blah_is_defined
static int blah = 0; // a different initial value
#endif

int bar() {
  // uses blah
}
```

tautschnig replied:
The call to `link_goto_model` will yield a warning in such a case (which I have just tested to make sure this really works as intended).

tautschnig committed:
C/C++ library: move free/delete-only definitions to library

With the support of building initialisers while adding the library we
can move library-only symbols to the library. This avoids polluting the
initialisation function for all cases that don't actually use
memory (de-)allocation library functions.

tautschnig committed:
C library: move pthreads related definitions to library

With the support of building initialisers while adding the library we
can move library-only symbols to the library. This avoids polluting the
initialisation function for all cases that don't actually use
pthreads-related library functions.

This removal of unused symbols required adding a symbol into the
array_of_bool_as_bitvec test, which previously implicitly relied on
those library-specific symbols in the patterns being tested for.
